tests/samples.ts

```diff
@@ -215,3 +139,118 @@ export const albMultiValHeadersRequest = (): ApplicationLoadBalancerRequestEvent
+      userArn: null,
+   };
+};
+
+export const makeAPIGatewayRequestContext = (params?: MakeAPIGatewayRequestEventContextParams): APIGatewayEventRequestContext => {
+   return {
+      accountId: '123456789012',
+      apiId: 'someapi',
+      authorizer: null,
+      httpMethod: params?.httpMethod ?? 'GET',
+      path: '/prd',
+      protocol: 'HTTP/1.1',
+      stage: 'prd',
+      requestId: 'a507736b-259e-11e9-8fcf-4f1f08c4591e',
+      requestTimeEpoch: 1548969891530,
+      resourceId: 'reas23acc',
+      identity: makeAPIGatewayRequestContextIdentity(),
+      resourcePath: '/',
+   };
+};
+
+export const makeAPIGatewayRequestEventHeaders =
+   (params?: MakeAPIGatewayRequestEventHeadersParams): APIGatewayRequestEvent['headers'] => {
+      return withDefault<APIGatewayRequestEvent['headers']>(
+         params,
+         {
+            Accept: '*/*',
+            'CloudFront-Forwarded-Proto': 'https',
+            'CloudFront-Is-Desktop-Viewer': 'true',
+            'CloudFront-Is-Mobile-Viewer': 'false',
+            'CloudFront-Is-SmartTV-Viewer': 'false',
+            'CloudFront-Is-Tablet-Viewer': 'false',
+            'CloudFront-Viewer-Country': 'US',
+            Host: 'b5gee6dacf.execute-api.us-east-1.amazonaws.com',
+            'User-Agent': 'curl/7.54.0',
+            Via: '2.0 4ee511e558a0400aa4b9c1d34d92af5a.cloudfront.net (CloudFront)',
+            'X-Amz-Cf-Id': 'xn-ohXlUAed-32bae2cfb7164fd690ffffb87d36b032==',
+            'X-Amzn-Trace-Id': 'Root=1-4b5398e2-a7fbe4f92f2e911013cba76b',
+            'X-Forwarded-For': '8.8.8.8, 2.3.4.5',
+            'X-Forwarded-Port': '443',
+            'X-Forwarded-Proto': 'https',
+            Referer: 'https://en.wikipedia.org/wiki/HTTP_referer',
+            Cookie: 'uid=abc; ga=1234; foo=bar; baz=foo%5Ba%5D; obj=j%3A%7B%22abc%22%3A123%7D; onechar=j; bad=j%3A%7Ba%7D',
+         }
+      );
+   };
+
+export const makeAPIGatewayRequestEventMultiValueHeader =
+   (params?: MakeAPIGatewayRequestEventMultiValueHeadersParams): APIGatewayRequestEvent['multiValueHeaders'] => {
+      return withDefault<APIGatewayRequestEvent['multiValueHeaders']>(
+         params,
+         {
+            Foo: [ 'bar', 'baz' ],
+            ...Object.fromEntries(
+               Object.entries(makeAPIGatewayRequestEventHeaders())
+                  .map(([ key, value ]): Array<string | Array<string | undefined>> => {
+                     return [ key, [ value ] ];
+                  })
+            ),
+         }
+      );
+   };
+
+export const makeAPIGatewayRequestEvent = (params?: MakeAPIGatewayRequestEventParams): APIGatewayRequestEvent => {
+   const httpMethod = withDefault(params?.httpMethod, 'GET'),
+         path = withDefault(params?.path, '/echo/asdf/a'),
+         headers = withDefault(params?.headers, makeAPIGatewayRequestEventHeaders());
+
+   const multiValueHeaders = withDefault(
+      params?.multiValueHeaders,
+      makeAPIGatewayRequestEventMultiValueHeader()
+   );
```

Comment:
With this approach we're putting the burden of ensuring `headers` and `multiValueHeaders` are in sync on the developer writing the tests. Do we have tests which need to override only `multiValueHeaders`? If so, what if `multiValueHeaders` is normally generated dynamically from `headers` but can have overrides provided via `multiValueHeaders`. If we go that route `MakeAPIGatewayRequestEventParams` will need a comment clearly explaining that `headers` should normally be used but `multiValueHeaders` can be used to selectively override multi-value header values.

Reply:
We currently don't have tests that override only multiValueHeaders. Some places replace them inside the test by creating a copy with _.extends and specifying only the needed header.

Reply:
@onebytegone If we implement it to generate `multiValueHeaders` from `headers`, how things should work when the developer only need to think about `multiValueHeaders`? If he tries changing it through the maker function params, using the `multiValueHeaders`, which is what interests him, he'll need to check implementation details of the function. Do you consider it a problem?